### DISCUSSION TITLE
Forward auth

### DISCUSSION THREAD

**Discussion by AndreKR** [zitadel/zitadel#5537] [Ideas]
Maybe it's already supported, but if not: I think it should be possible to use Zitadel for _forward auth_ authentication, for example with [Traefik](https://doc.traefik.io/traefik/middlewares/http/forwardauth/), [nginx](http://nginx.org/en/docs/http/ngx_http_auth_request_module.html) or [Caddy](https://caddyserver.com/docs/caddyfile/directives/forward_auth).

**Comment by fabius** [comment]
It's possible to hook up proxies like [oauth2-proxy](https://github.com/oauth2-proxy/oauth2-proxy) to do just that. It is sparsely documented [over here in the Zitadel docs](https://zitadel.com/docs/self-hosting/manage/reverseproxy/reverse_proxy). It works though!

**Reply by fforootd** [reply]
Yeah some proxies work with OIDC/Oauth our docs need some love on that end though 😁

**Reply by fabius** [reply]
@fforootd  I'd be happy to contribute a bit more context / an example for the oauth2-proxy configuration. Unless someone beats me to it of course. Is this being worked on at the moment or are you taking a PR?

**Reply by fforootd** [reply]
I would love to review a PR. The[ initial docs](https://zitadel.com/docs/examples/identity-proxy/oauth2-proxy) was done by me.

Btw. the [flat roles claim we now have ](https://zitadel.com/blog/custom-claims) should work well along the "groups claim" concept that the oauth proxy has

**Reply by AndreKR** [reply]
Are we still talking about *forward* auth (webserver asks for authentication status before serving the resource) or are we talking about an auth proxy (traffic goes through the auth proxy)?

**Reply by fabius** [reply]
We are talking about *forward auth*. However, your reverse proxy will not delegate the decision making directly to Zitadel but to the oauth2-proxy which in turn will reach out to Zitadel. So there's one more moving part, but the end result is exactly what you are after

**Reply by AndreKR** [reply]
Ah, now I found it.

oauth2-proxy has an `/auth` endpoint that can be used for forward auth. There are instructions for [nginx](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview#configuring-for-use-with-the-nginx-auth_request-directive) and [Traefik](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview#configuring-for-use-with-the-traefik-v2-forwardauth-middleware).

I'll try it out when I find the time. I have tried [traefik-forward-auth](https://github.com/thomseddon/traefik-forward-auth) in the past but it had [an issue](https://github.com/thomseddon/traefik-forward-auth/issues/137) with path prefixes.

**Reply by arthurgeek** [reply]
hey @AndreKR! 👋

have you got this working?

**Reply by AndreKR** [reply]
Still haven't tried it yet. 😄

**Reply by kekalainen** [reply]
@arthurgeek I can confirm that oauth2-proxy `v7.4.0` works with Zitadel `v2.30.0` and Traefik `v3.0.0-beta3`'s ForwardAuth middleware (tested with the [`api@internal`](https://doc.traefik.io/traefik/operations/dashboard/#secure-mode) service on Docker `v24.0.4` swarm mode).

**Reply by Jozefiel** [reply]
hi @kekalainen .
Can you post an example of configuration?
It can be a great source of knowledge. 

Thanks

**Reply by kekalainen** [reply]
Hey @Jozefiel, [here's](https://gist.github.com/bc8168a4209be350ea42f42d497bd758) an example based on the swarm mode service configuration I used.

**Reply by tafaust** [reply]
Hey @kekalainen, so I'd need to forwardauth from traefik to oauth2-proxy which handles oidc auth with zitadel? Can't it be done without oauth2-proxy? Thanks for your example!

**Reply by LeVraiRoiDHyrule** [reply]
Hi, I'm very interested into what you did. I am trying to do the same thing with Caddy. I've set caddy this way:
```
radarr.example.com {
        @except not path /api*
        reverse_proxy radarr:7878
        reverse_proxy /oauth2/* oauth2-radarr:4180
        forward_auth @except oauth2-radarr:4180 {
                copy_headers Authorization
                uri /oauth2/auth
        }
}
```

And Oauth2-proxy this way:
```
      - OAUTH2_PROXY_HTTP_ADDRESS=0.0.0.0:4180
      - OAUTH2_PROXY_EMAIL_DOMAINS=*
      - OAUTH2_PROXY_PROVIDER=oidc
      - OAUTH2_PROXY_REDIRECT_URL=https://radarr.${DOMAIN}/oauth2/callback
      - OAUTH2_PROXY_OIDC_ISSUER_URL=https://auth.${DOMAIN}
      - OAUTH2_PROXY_WHITELIST_DOMAIN=.${DOMAIN}
      - OAUTH2_PROXY_SKIP_PROVIDER_BUTTON=true
      - OAUTH2_PROXY_CLIENT_ID=${RADARR_CLIENT_ID}
      - OAUTH2_PROXY_CLIENT_SECRET=dummy #${RADARR_CLIENT_SECRET}
      - OAUTH2_PROXY_CODE_CHALLENGE_METHOD=S256
      - OAUTH2_PROXY_OIDC_GROUPS_CLAIM=groups
      - OAUTH2_PROXY_SCOPE="openid profile email groups"
      - OAUTH2_PROXY_COOKIE_SECRET=${OAUTH2_PROXY_COOKIE_SECRET}
```
This is based on your config, I just changed code challenge method to PKCE as it is recommended by Zitadel.

But I am having the error "Unauthorized" from oauth2-proxy, I don't know more about that (I am also having this error when not using PKCE). Would you know what could cause this ?

Do I need to have a subdomain for Oauth2-proxy ?
How did you set your app in Zitadel ? Did you do anything besides creating the app and using the clientid/secret Zitadel gave you ?

Thanks in advance for any answer and have a great day

**Comment by Kryszn0** [comment]
@kekalainen would you be able to share how you setup your application within Zitadel to work with this setup? I have managed to get Zitadel working work with OAuth2 Proxy and NGINX-Proxy-Manager, but can't seem to figure out how to have Zitadel authorize access to the original URL I am trying to protect. 

To clarify, I've attached an image of the general flow being used and below is what is happening/my issue:
1. Navigate to app.domain.com
2. Redirected to auth.domain.com to login to Zitadel instance

Expected result after successful login:
3. Redirected to app.domain.com

What is happening:
3. Redirected to a URL that generally follows this: auth.domain.com/oauth2/auth/code=${RANDOMSTRING}%https%app.domain.com%     
The webpage itself is a black page that simply states 'Unauthorized' on the top left. 

I believe my problem is simply I haven't correctly told Zitadel that the user I am logging in as is allowed to access app.domain.com . However, I can't seem to figure out how to do this. I'm hoping I can use this setup to set Zitadel as a front for any of the apps I am self-hosting (AdGuardHome, FileBrowser, etc.). Integrating Zitadel's userbase with the logins of the apps themselves is a later project/task. 

![proxy](https://github.com/zitadel/zitadel/assets/113187314/91af6a9b-3b85-4990-9221-f0e5afe76fc8)

**Reply by muhlemmer** [reply]
Redirect URI configuration is documented here: https://zitadel.com/docs/guides/manage/console/applications#redirect-uris

If you use a Oauth2 proxy, I can imagine it doesn't support OpenID Connect discovery, so you'll need to somehow configure the authorization server [endpoints](https://zitadel.com/docs/apis/openidoauth/endpoints) manually. If you get an `Unauthorized` back it might be that the `introspection` endpoint is not configured correctly.

**Reply by LeVraiRoiDHyrule** [reply]
Hi,
I am a begginer with Zitadel, I was using Authentik before that integrates a proxy that I used for forward_auth with Caddy. Did someone suceed to set up forward_auth with caddy and oauth2-proxy ? I can't find any example and I don't know where to begin to set this up.

**Reply by ad-on-is** [reply]
hey @LeVraiRoiDHyrule Have you found any solution to this? I'm struggling myself with it.

Running caddy, zitadel and oauth2-proxy. But no matter what config I try, it doesn't work. The one I use right now, authenticates me successfully, but does not redirect properly to app.example.com, oauth2proxy throws "Rejecting invalid redirect" errors.

**Reply by LeVraiRoiDHyrule** [reply]
I did not do another attemps on Oauth2-proxy + Zitadel, I am currently focused on apps natively supporting OpenID. But I will still need it for some apps on my homelab so I will give another try. However, I am no longer using caddy and switched a nginx-based reverse proxy (bunkerweb). So I won't be able to help for caddy x oauth2-proxy. But I have seen on the oauth2-proxy github issues examples from people using it with caddy, I will try to find that. oauth2-proxy x zitadel seems easier thogh, I succeeded to connect it using PKCE.

**Reply by ad-on-is** [reply]
@LeVraiRoiDHyrule I got it working!!!

I switched from using ENV-variables to a cfg-file. In case you want to give it a shot... Here's my setup.

```
provider = "oidc"
user_id_claim = "sub" #uses the subject as ID instead of the email
provider_display_name = "ZITADEL"
redirect_url = "https://oauth2proxy.example.com/oauth2/callback"
oidc_issuer_url = "https://zitadel.example.com"
upstreams = []
email_domains = [
    "*"
]
client_id = "xxxx"
client_secret = "xxxxx"
code_challenge_method = "S256"
pass_access_token = true
cookie_secret = "xxxxx"
skip_provider_button = true
cookie_secure = false #localdev only false
http_address = "0.0.0.0:4180" #localdev only
whitelist_domains = ".example.com"
cookie_domains = ".example.com"
reverse_proxy = true
set_xauthrequest = true
```

Caddyfile

```
(secureoauth) {
  reverse_proxy /oauth2/* oauth2proxy:4180
  handle {
  forward_auth oauth2proxy:4180 {
    uri /oauth2/auth
    copy_headers Authorization X-Auth-Request-User X-Auth-Request-Email
    @bad status 4xx
    handle_response @bad {
      redir https://oauth2proxy.example.com/oauth2/start?rd={scheme}://{host}{path}?{query}
    }
  }
  }
}

@bin host bin.example.com
handle @bin {
  import secureoauth
  reverse_proxy microbin:8080
}
```

```yml
oauth2proxy:
    image: quay.io/oauth2-proxy/oauth2-proxy
    container_name: oauth2proxy
    user: 1000:1000
    restart: always
    volumes:
      - /srv/docker/oauth2proxy/oauth2-proxy.cfg:/etc/oauth2-proxy.cfg
    depends_on:
      - zitadel
    command: ["oauth2proxy", "--config=/etc/oauth2-proxy.cfg"]
```

**Reply by JonnyMe** [reply]
In case it can be helpful, here is my current integration (Pulumi + k8s):
 https://github.com/JonnyMe/traefik-oauth2proxy-zitadel-homarr-integration

**Reply by fforootd** [reply]
that's cool!

thank you for sharing this.

**Reply by JonnyMe** [reply]
I believe giving back is the core to open source. BTW I love the automation level Zitadel offers, thank you for your great work! 🤗

**Reply by divStar** [reply]
I am not able to set up `oauth2-proxy` with `zitadel` and `traefik` (v3) and `whoami`, which is expected to use `oauth2-proxy`, which in turn should redirect to Zitadel.
I am basically stuck on whether I set `- "traefik.http.middlewares.oauth2-auth.forwardauth.address=http://oauth2-proxy:4180/"` or `- "traefik.http.middlewares.oauth2-auth.forwardauth.address=http://oauth2-proxy:4180/oauth2/auth"`.

If I use the former, I am redirected to Zitadel, but after the login I get a `404` always.
If I use the latter, I am not redirected, but simply see a `401 Unauthorized` page. IF I am already logged-in though, the website shows the expected content.

I am really out of ideas. Authelia indeed makes this so simple even someone like me, who isn't too much into OIDC set up etc. can do it in seconds, but Zitadel is so much better in terms of management in every way. Can someone give me a minimal example of traefik v3, zitadel, oauth2-proxy and whoami (that uses oauth2-proxy/zitadel)?
I can provide my `docker-compose.yml` files if need be, though I'd have to redact the secrets etc. first.
